### PR TITLE
Deprecate `limit_amplitude` constructor argument

### LINKED ISSUE #7932: Pulse `limit_amplitude` is confusing
### What should we add?

The class variable `limit_amplitude` in `Pulse` base class controls validation of max pulse amplitude. According to how test is written, likely this is designed as a global variable. 

https://github.com/Qiskit/qiskit-terra/blob/a735680f69f7a2e2757dddbedf09f6d7c6a5e692/test/python/pulse/test_pulse_lib.py#L85-L93

However, the every pulse subclass constructor takes `limit_amplitude`, and this may confuse users because it looks like `limit_amplitude` is an instance variable.

https://github.com/Qiskit/qiskit-terra/blob/a735680f69f7a2e2757dddbedf09f6d7c6a5e692/qiskit/pulse/library/parametric_pulses.py#L103-L125

```python
with pulse.build(backend) as sched:
    pulse.play(Gaussian(duration=160, amp=1.6, sigma=40, limit_amplitude=False), d0)
    pulse.play(Gaussian(duration=160, amp=0.8, sigma=40, limit_amplitude=True), d0)
```

If something like this is supported, `limit_amplitude` should be really instance variable, i.e. here the parameter validation happens only in the constructor, so setting `limit_amplitude` for the first pulse only affects the first pulse, if this variable is specified by every class instantiation.

I think `limit_amplitude` should be removed at least from all pulse subclass constructor. It would be great Qiskit offers some mechanism to manage _preference_ of users, rather than directly attaching such preference to class.

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary

This PR deprecates `limit_amplitude` which is a constructor argument of pulse subclass. This option is confusing because user may unintentionally update class variable, or sometime instance variable.

### Details and comments

Fix #7932

### CODE REVIEW

**Comment by jakelishman** [comment]
I don't actually use the code, so I'm not the best-placed person to be involved in the decision, but certainly from the outside it _feels_ like instance-level control of limiting the amplitude is a useful thing to have.  Setting the class-level attribute affects already-instantiated classes, which is (to me) very surprising to be the suggested path.

**Comment by nkanazawa1989** [comment]
From control electronics point of view, instance level amplitude limit is bit unrealistic. Historically this control parameter is introduced to manage pulse amplitude in SI value, which is based upon the request from the community. 

Usually, when we design the controller, we decide how we control the pulse amplitude. In IBM provider, amplitude indicates gradation of quantized amplitude (i.e. level of AWG output) in the digital world which will be converted into analog signal by mixing with local oscillator or LO signal (the power of the oscillator is not controllable from Qiskit). This means amplitude beyond 1.0 should not exist because we cannot output power exceeding LO power, i.e. 1.0 indicates 100% of LO power.

As such, we cannot mix different amplitude representations in the same pulse channel because of physically implemented hardware. This means `limit_amplitude` is the control parameter of the `Channel` object rather than `Pulse` instance. However, since pulse instance is not aware of channel being associated with, thus it cannot retrieve limit amplitude policy from channel at instantiation. 

The assumption behind the PR is -- if one design a control electronics that employs SI amplitude, then one will not implement  the whole control system with channels with different amplitude representations. Then, `limit_amplitude` is something that is set before one start building pulse programs.

**Comment by nkanazawa1989** [comment]
This makes me think we can delay the validation of amplitude or entire pulse parameters. If `Play(pulse, channel)` instruction instance does the validation, we can get amplitude limit policy from the cannel, which is physically makes sense.

File: test/python/pulse/test_pulse_lib.py [hunk 1]
```diff
@@ -282,37 +275,44 @@ def test_hash_generation(self):
 
     def test_gaussian_limit_amplitude(self):
         """Test that the check for amplitude less than or equal to 1 can be disabled."""
-        waveform = Gaussian(duration=100, sigma=1.0, amp=1.1 + 0.8j, limit_amplitude=False)
-        self.assertGreater(np.abs(waveform.amp), 1.0)
-
         with self.assertRaises(PulseError):
-            Gaussian(duration=100, sigma=1.0, amp=1.1 + 0.8j, limit_amplitude=True)
+            GaussianSquare(duration=100, sigma=1.0, amp=1.1 + 0.8j, width=10)
```
Comment (eggerdj) [inline test/python/pulse/test_pulse_lib.py:279]: Why the change to GaussianSqure?

File: qiskit/pulse/library/pulse.py [hunk 2]
```diff
@@ -47,6 +48,12 @@ def __init__(
         self.duration = duration
         self.name = name
         if limit_amplitude is not None:
+            warnings.warn(
+                "Instantiating a pulse instance with 'limit_amplitude' has been deprecated. "
+                f"Please directly update class variable '{self.__class__.__name__}.limit_amplitude.'",
+                DeprecationWarning,
+                stacklevel=4,
+            )
```
Comment (eggerdj) [inline qiskit/pulse/library/pulse.py:56]: Would it not be more flexible to deprecate the class variable instead and do this validation from one instance to the next? With this PR you either validate all your pulses or none of them. On the other hand, I'm not sure if we need the flexibility.

File: test/python/pulse/test_pulse_lib.py (hunk 1 above)
Comment (nkanazawa1989) [reply test/python/pulse/test_pulse_lib.py:279]: Good catch this is a typo.

File: qiskit/pulse/library/pulse.py (hunk 2 above)
Comment (nkanazawa1989) [reply qiskit/pulse/library/pulse.py:56]: Yeah that's the point I'd like to hear your thought. I think instance wise validation is not practical from viewpoint of hardware, because we cannot dynamically change representation in the same channel. I `limit_amplitude` is more like a instance variable of channel rather than pulse.
https://github.com/Qiskit/qiskit-terra/pull/7948#issuecomment-1102019646

**Comment by taalexander** [comment]
>I don't actually use the code, so I'm not the best-placed person to be involved in the decision, but certainly from the outside, it feels like instance-level control of limiting the amplitude is a useful thing to have. Setting the class-level attribute affects already-instantiated classes, which is (to me) very surprising to be the suggested path.

I tend to agree with @jakelishman, why are we switching this to a class variable? What about the situation where we might have two users (say one a simulator, the other a real-backend) in the same process? Or the situation where we don't want to limit to 1 due to an optimization routine, that we later rewrite all pulses that exceed 1 to be 1?

**Comment by nkanazawa1989** [comment]
The issue of current implementation is delegation that @jakelishman pointed out. I'm currently cleaning up the data structure of pulse schedule before starting the QPY serialization (i.e. in circuit QPY we have to maintain tons of branching logic to cover many variants of instruction data, so I'm trying to make pulse data structure bit more strict). In this sense, I find the `limit_amplitude` is quite problematic, since it's difficult to write clean code to infer situation of delegation. 

Another simple option would be just turning it into instance variable, but this means a coder must set `limit_amplitude=True` to every pulse instance which is very cumbersome. From hardware point of view I think this option belongs to `PulseChannel` rather than pulse, but as you mentioned
> What about the situation where we might have two users (say one a simulator, the other a real-backend) in the same process?
this could be argument of `.run`.

**Comment by jakelishman** [comment]
If it's just the delegation and QPY serialisation we're primarily worried about, then the instance-level attributes will be easiest to standardise around rather than the class-level attributes - QPY currently deals with objects that can be serialised with no global state.

Perhaps a more sensible state is if the class-level attribute is used during object instantiation as the default, but there's no delegation after that (e.g. something like `self.limit_amplitude = type(self).limit_amplitude if limit_amplitude is None else limit_amplitude`).  That still lets you set things on a larger scale (though personally I find class variables aren't much better than true globals - it's asking for problems in multithreading), but stops the issue of later changes affecting already created objects.  QPY wouldn't need to worry any more, because it could just restore the state it found.

That doesn't solve the problem you mentioned about the amplitude limiting being more a property of a `Channel` than a `Pulse`, but I'm not sure that's completely solvable without a larger re-write.  At any rate, removing the instance variables and making it a global only means we can't simultaneously describe one `Channel` with limiting and one without.

**Comment by nkanazawa1989** [comment]
In #7821 limit_amplitude is turned into an instance variable (`self._limit_amplitude`). This is sufficient for QPY serialization.

**Comment by wshanks** [comment]
I'll just throw my opinion in here in case the larger design gets revisited later.

I don't see much of a use case for setting `limit_amplitude` on individual pulses. The current behavior is just to raise an exception if an amplitude over the limit is detected. For running on IBM hardware at least, any amplitude over the limit will cause the provider to give back an error. Getting the error in terra saves you sitting through the queue, so you want to check every pulse. For a pulse simulator, there is probably no limit (other than maximum allowed floating point value) so you don't want to check any pulse.

@taalexander imagined a more interesting case where amplitudes could be rescaled. An interesting variant would be a provider that silently scaled down amplitudes to the max values. Perhaps you would not care that some pulses were getting scaled but would want to make sure others were exactly as specified and so marking individual pulses would make sense?

I agree that basing the behavior off a class level variable is not good. @nkanazawa1989 has talked about adding a pass manager for pulse schedules. I think amplitude limit checking would fit better as a validation pass than something that is checked at class  instantation and parameter assignment. Perhaps such a pass could take filtering arguments for choosing which pulses to limit or not so that this does not need to be set on each individual pulse.

**Comment by taalexander** [comment]
I think you're right about how to approach amplitude verification @wshanks it should also have performance improvements
